### DISCUSSION TITLE
Access INP file without running a simulation

### DISCUSSION THREAD

**Discussion by JoaoBarreiro** [pyswmm/swmmio#201] [Q&A] [answered]
Hello,
I am implementing an analysis after running the simulation, for which I need some of data INP data. I was implementing something like the code below. Correct me if I'm wrong, but by doing this I am starting the simulation but not going through its timesteps. How can I access the INP File data without starting/running  the simulation?

def getElementsProperties(InpFilePath):
    
    with Simulation(InpFilePath) as sim:
        NodesDict = {}
        for node in Nodes(sim):
            new_node = {}
            if node.is_junction:
                new_node["InvertElevation"] = node.invert_elevation
                new_node["FullDepth"] = node.full_depth    
            NodesDict[node.nodeid] = new_node
                
        LinksDict = {}
        for link in Links(sim):
            if link.is_conduit:
                new_link = {}
                new_link["Nodes"] = link.connections   # tuple (“UpstreamNodeID”,”DownstreamNodeID”)
                new_link["InletOffset"] = link.inlet_offset
                new_link["OutletOffset"] = link.outlet_offset
            LinksDict[link.linkid] = new_link
    
    return NodesDict, LinksDict

**Comment by bemcdonnell** [comment]
@JoaoBarreiro, I think you should use swmmio for accessing info. You are correct, pyswmm.Simulation “opens” a simulation (which will overwrite your RPT/OUT). I’m guessing that is the problem you’re experiencing.  Swmmio and pyswmm are friends at the moment and will soon be coupled together more tightly.  I’m going to transfer this discussion to swmmio

**Reply by JoaoBarreiro** [reply]
@bemcdonnell thanks for the reply. That's what I thought. I wasn't aware of the existence of swmmio, and I will explore this.
Thanks

**Reply by bemcdonnell** [reply]
@aerispaha care to provide a simple example? :-)

**Comment by aerispaha** [comment, marked as answer]
Hi @JoaoBarreiro, as @bemcdonnell suggested, you can used swmmio to fetch data from the INP file without needing to run the simulation. Here is a basic example:

```python 
import swmmio

# instantiate a swmmio Model object with your inp file
model = swmmio.Model('path/to/model.inp') 

# access an aggregation of inp data related to model nodes, using the higher level api
nodes = model.nodes.dataframe 

# similarly, access an aggregation of data related to model links
links = model.links.dataframe 
```

Here, `nodes` will be Pandas DataFrames with columns for InvertElev, SurchargeDepth, and many others. Similarly, `links` will be a DataFrame containing InletNode, OutletNode, Geom1 etc.

Finally, if you need data from the INP that isn't provided in the higher level API via `model.nodes` and `model.links`, you can access individual sections in the INP file using the lower-level `model.inp` API. For example, we can fetch the [DWF] section of an INP file (if it exists) as a dataframe like this: 

```python
model.inp.dwf 
```
More documentation on this can be found over [here](https://swmmio.readthedocs.io/en/v0.6.7/swmmio.html#swmmio.core.inp). 

Does that answer your question?
